### PR TITLE
cmake Python changes

### LINKED ISSUE #2730: Python modifications for cmake
Part of #2689, but wanted a clean page to contemplate how best to adapt the API / Python executables to the modified filesystem structure & installation process for `cmake`.

-----

1.  Some of the logic in `bin/mrtrix3.py` is tailored for running Python executables that are part of external MRtrix3 projects. Relevant documentation [here](https://mrtrix.readthedocs.io/en/3.0.4/tips_and_tricks/external_modules.html#handling-python-commands). There, the logic proceeds something like: 1) If a symbolic link to the main repo `bin/mrtrix3.py` has been made within the external project, that can be used to find the API. 2) Often external projects are built by constructing a softlink to the `build` script in the main repo and running that directly; that softlink can therefore be used to find the API. Use of `PYTHONPATH` is a fallback, but if used would kick in before any of this logic.

    This is being made redundant; indeed it's unclear at this point in time exactly what external projects will look like after the cmake changes.

    **Suggestion**: Remove all support for external modules within the code. `PYTHONPATH` will remain a viable solution. At the point at which we address compilation of external C++ binaries we will also address external Python executables.

2.  It is intended to be possible for a user to define a symbolic link to the executable, and attempting to run that executable will still find the Python API based on a relative path from the resolved symbolic link.

3.  The `run` module includes logic to ensure that if an *MRtrix3* executable is invoked, it is the coinciding version of that executable that is invoked rather than the first that appears in `PATH`, and also there are certain command-line options that may be added to `run.command()` invocations based on the availability of "standard" *MRtrix3* command-line options.

    The logic behind this may need to be revised. Currently, it just looks at the contents of `../../bin/`. So if the software is installed into a location that is shared with other software, eg. `/usr/local/`, that logic will catch a large set of non-MRtrix3 commands. 

    **Suggestion**: The build stage of cmake could collect a list of *MRtrix3* executables, and write it to a text file within `lib/mrtrix3/` in the build directory. `lib/mrtrix3/__init__.py` would then read from this file.

4.  It was hypothesized that it would be nice to be able to:

    1.  Run *MRtrix3* Python scripts from either an installed location or the build directory, and for point 3 above to still be satisfied.
    
    2.  Modify the contents of Python executables within the source directory, and have those changes reflected in the build directory without having to re-run the build step of cmake.
    
        However on reflection it may be tricky to support *both* this feature *and* point 2 above. It would be treating softlinks in different ways, and also ideally this modify-without-build feature would apply to the API and script algorithms also.

    **Suggestions**: For point i., as long as the directory structure is consistent between build and install, and there isn't any softlink trickery, it should work out of the box. For point ii., just rely on cmake build to copy any Python updates made in the source code into the build directory. 

5.  Already [mentioned](https://github.com/MRtrix3/mrtrix3/pull/2689#issuecomment-1748012497) in #2689, but relevant here also: Placement of the script algorithms.

    Before the `cmake` changes, there are executable Python scripts in `bin/`, APIs in `lib/mrtrix3/`, and algorithm source code in `lib/mrtrix3/<exename>/*.py`. Thus far for cmake this has all just been moved into new directory `python/*` but the relative structure is unchanged.

    **Suggestions**: There's multiple options here. One I'm increasingly liking as suggested by @aaroncapon is to move the sub-directories of `python/lib/mrtrix3/` (ie. the script algorithms) into `python/src/mrtrix3/`, and within the build / installation directory populate `src/mrtrix3/` with these contents. @daljit46 do you see any issue with the build / installation containing a `src/mrtrix3/` directory?

### PR DESCRIPTION
**Pull request by Lestropie**
Closes #2730.

My third attempt at Python refactoring following `cmake` transition (#2689), after #2737 and #2741. Done a decent amount of thinking about it after it was last spoken about during a dev team meeting. I'm happy with this one and want to proceed. Commands all seem to execute as expected, from both build and install directories. For existing Python commands, requisite change is minimal; just deletion of the "`import mrtrix3; mrtrix3.execute()`" at the tail of the file.

Lots of details in commit message of 93f3741c68b1fe8a350b05a155aa7272ff2775e2.

@daljit46: Handballing to you. You have carte blanche to change as you see fit; whether to resolve my `cmake` naivety, or anything in this proposal that would be completely prohibitive if trying to expand the concept to external project support. But in the absence of any major criticisms from yourself or any other @MRtrix3/mrtrix3-devs, hoping this gets me out of your way to proceed with #2901.

### CODE REVIEW

**Comment by Lestropie** [comment]
- [ ] For Python-API-based testing tools, need to have `cmake` generate executables just as is done for Python commands.

**Comment by daljit46** [comment]
Looks mostly fine to me at a first glance. 

> API files reside in python/mrtrix3/ (still in a sub-directory called "mrtrix3/" so that IDEs can reasonably identify them, but no need to be in a "lib/" sub-directory in source form). 

I don't quite the reasoning behind this. How does this help IDEs?

**Comment by Lestropie** [comment]
> I don't quite the reasoning behind this. How does this help IDEs?

Not actually tested, but I assume that if an IDE sees `from mrtrix3 import app`, and you try to hotkey from function invocation to declaration, that is far more likely to succeed if file "`app.py`" resides in a directory called "`mrtrix3/`"; hence putting those files in `python/mrtrix3/` rather than `python/`.

**Comment by daljit46** [comment]
I see, that makes sense. Just tested this on VS Code and it behaves as you predicted.

**Comment by daljit46** [comment]
I think it would make sense for `cmake/MakePythonExecutable.cmake` to accept a `DESTINATION_PATH` argument instead of `BUILDDIR`. This way we can use the logic in the script for testing tools tool.

**Comment by daljit46** [comment]
I'm a little unsure on how to handle `testing_python_cli`. Currently, it resides in `testing/tools`, but I think this is not the appropriate location for it. The command is not even a "tool", in the sense that it's not really used by multiple scripts/executables to verify the correctness of input/outputs. Instead, it's just a "fake" python command used to test that the CLI works as intended.

Furthermore, currently we create the testing tools in `${PROJECT_BINARY_DIR}/testing/tools` (which is added to PATH for any bash test invocation), but perhaps it makes sense to set their build output directory to `${PROJECT_BINARY_DIR}/bin`. This has the benefit that whatever mechanism we use for standard Python commands should, in theory, also for executing a testing tool.

**Comment by Lestropie** [comment]
RE `testing_python_cli`:

-   If it were the case that our `CMakeLists` didn't involve globbing, then that file could happily reside in `python/mrtrix3/commands/`, and just be omitted from the list of generated executables if not building tests. I'm not sure if this case is sufficient to make that transition, just making the observation.
-   While it's true that the number of tests that invoke this command is only 1, I'm not sure if that distinction is adequate to motivate conceptualising it differently to the other testing tools. It is nevertheless still "a tool that is invoked exclusively by tests".

RE the build path for testing tools, personally I find it a little unusual to look inside of the `bin/` directory and see commands in there that are exclusively for testing. Eg. There's a command called `to`. So if anything I'd prefer to be moving those "unit test" test commands out of `bin/` and into `testing/tools/`, rather than moving anything in the opposite direction.

**Comment by daljit46** [comment]
I have made some changes to fix the running of unit tests. `cmake/MakePythonExecutable.cmake` now accepts an `OUTPUT_DIR` directory where the output executable will be generated, and also takes an optional list `EXTRA_PATH_DIRS` that can be used to specify additional paths to be prepended to `sys.path`.
In `testing/tools/CMakeLists.txt`, now there is a new function `add_python_tool` which can generate a Python test executable that behaves like a standard MRtrix3 python command (using `MakePythonExecutable.cmake`). This is a slightly cumbersome approach, but it does seem to work.
@Lestropie any feedback welcome.

File: cmake/MakePythonExecutable.cmake [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    endforeach()
+endif()
+
+string(APPEND BINPATH_CONTENTS "from mrtrix3.app import _execute\n")
+
+# Three possible interfaces:
+#   1. Standalone file residing in commands/
+#   2. File stored in location commands/<cmdname>/<cmdname>.py, which will contain usage() and execute() functions
+#   3. Two files stored at commands/<cmdname>/usage.py and commands/<cmdname>/execute.py, defining the two corresponding functions
+# TODO Port population_template to type 3; both for readability and to ensure that it works
```
Comment (Lestropie) [inline cmake/MakePythonExecutable.cmake:33]:
```suggestion
```

**Comment by Lestropie** [comment]
Yep, that solution for `testing_python_cli` looks exactly like what I had in mind.

Python command documentation generation is failing on Linux CI:
`find: '/home/runner/work/mrtrix3/mrtrix3/python/bin/': No such file or directory`
Is this maybe a cache problem? The workflow makes no reference to that deprecated location.

Once that's resolved I'm happy for this to be merged.

**Comment by Lestropie** [comment]
Spoke too soon; one of the changes you made to the cmake-generated Python executables to facilitate `testing_python_cli` broke the execution of some of the Python commands.

Full details in 3fce6484a071acd6ea703b05bdf9e21064153629, but what I chose to do is put `testing_python_cli.py` into `${PROJECT_BINARY_DIR}/lib/mrtrix3/commands/` alongside the other command source code, and generate the executable in `${PROJECT_BINARY_DIR}/bin`; that way the same logic for the executable file generation applies.

See if you're happy with that. Assuming there isn't anything outstanding still broken, I'm happy to proceed with merging.

**Pull request merged**